test/unit/collector/serverless.test.js

```diff
@@ -97,192 +99,185 @@ tap.test('ServerlessCollector API', (t) => {
+      t.beforeEach(beforeEach)
+      t.afterEach(afterEach)
+
+      await t.test(`adds ${key} to the payload object`, (t) => {
+        const { api } = t.nr
         const eventData = { type: key }
         api.send(key, eventData, () => {
-          t.same(api.payload[key], eventData)
-          t.end()
+          assert.deepStrictEqual(api.payload[key], eventData)
         })
       })
 
-      t.test(`does not add ${key} to the payload object when disabled`, (t) => {
-        api.enabled = false
+      await t.test(`does not add ${key} to the payload object when disabled`, (t) => {
+        const { api } = t.nr
         const eventData = { type: key }
+        api.enabled = false
         api.send(key, eventData, () => {
-          t.same(api.payload[key], null)
-          t.end()
+          assert.equal(api.payload[key], null)
         })
       })
     })
-  })
-
-  t.test('#flushPayloadSync', (t) => {
-    t.autoend()
+  }
 
-    t.beforeEach(beforeTest)
-    t.afterEach(afterTest)
+  await t.test('#flushPayloadSync', async (t) => {
+    t.beforeEach(beforeEach)
+    t.afterEach(afterEach)
 
-    t.test('should base64 encode the gzipped payload synchronously', (t) => {
-      const testPayload = {
-        someKey: 'someValue',
-        buyOne: 'getOne'
-      }
+    await t.test('should base64 encode the gzipped payload synchronously', (t) => {
+      const { api } = t.nr
+      const testPayload = { someKey: 'someValue', buyOne: 'getOne' }
       api.payload = testPayload
-      const oldDoFlush = api.constructor.prototype._doFlush
+
+      let flushed = false
       api._doFlush = function testFlush(data) {
         const decoded = JSON.parse(zlib.gunzipSync(Buffer.from(data, 'base64')))
-        t.ok(decoded.metadata)
-        t.ok(decoded.data)
-        t.same(decoded.data, testPayload)
+        assert.notEqual(decoded.metadata, undefined)
+        assert.notEqual(decoded.data, undefined)
+        assert.deepStrictEqual(decoded.data, testPayload)
+        flushed = true
       }
       api.flushPayloadSync()
-      t.equal(Object.keys(api.payload).length, 0)
-      api.constructor.prototype._doFlush = oldDoFlush
-
-      t.end()
+      assert.equal(Object.keys(api.payload).length, 0)
+      assert.equal(flushed, true)
     })
   })
 
-  t.test('#flushPayload', (t) => {
-    t.autoend()
-
-    let outputSpy = null
-
-    t.beforeEach(() => {
-      // We're using NEWRELIC_PIPE_PATH to output to /dev/null so
-      // let's check that we are writing to the device.
-      outputSpy = sinon.spy(fs, 'writeFileSync')
-
-      beforeTest()
+  await t.test('#flushPayload', async (t) => {
+    t.beforeEach(async (ctx) => {
+      await beforeEach(ctx)
+
+      ctx.nr.writeSync = fs.writeFileSync
+      ctx.nr.outFile = null
+      ctx.nr.outData = null
+      fs.writeFileSync = (dest, payload) => {
+        ctx.nr.outFile = dest
+        ctx.nr.outData = JSON.parse(payload)
+        ctx.nr.writeSync(dest, payload)
+      }
```

Comment:
If there is any objection to my reducing the usage of `sinon` across these tests, I assume it will primarily be here (all of the other usages were very minimal to begin with). My justification is that the patching here is very clear to me, whereas with the original "spy" implementation I was very confused as to what `outputSpy.args[0][1]` even referred to. I had to set a few breakpoints to inspect what the actual code is doing in order to learn what this test was attempting to read.

I really don't think obscuring the work behind 3rd party DSL APIs is worth the cost. But if the team thinks otherwise, I can revert.

Reply:
I think that's just a personal preference but it's fine. I'll just let you know I def prefer the use of sinon over manually stubbing/spying